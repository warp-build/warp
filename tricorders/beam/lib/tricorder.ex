defmodule Tricorder do
  @moduledoc """
  Documentation for `Tricorder`.
  """

  @doc """
  Hello world.

  ## Examples

      iex> Tricorder.hello()
      :world

  """
  def hello do
    :world
  end
end
