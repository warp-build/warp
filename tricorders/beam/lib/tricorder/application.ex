defmodule Tricorder.Application do
  @moduledoc false

  use Application
  require Logger

  @impl true
  def start(_type, _args) do
    :code.delete(Tricorder.Project)
    :code.purge(Tricorder.Project)

    children =
      if Application.get_env(:tricorder, :start_server, false) do
        port = Application.get_env(:tricorder, :grpc_port) |> String.to_integer()

        [
          {GRPC.Server.Supervisor, {Tricorder.Grpc.Endpoint, port}},
          %{
            id: Tricorder.Deps.Server,
            start: {Tricorder.Deps.Server, :start_link, ["."]}
          }
        ]
      else
        []
      end

    opts = [strategy: :one_for_one, name: Tricorder.Supervisor]

    Supervisor.start_link(children, opts)
  end
end
