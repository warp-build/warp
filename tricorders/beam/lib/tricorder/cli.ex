defmodule Tricorder.CLI do
  require Logger

  def main(["start", port, workspace_root | _]) do
    Logger.info("Starting Tricorder application on port #{port}")

    Application.put_env(:tricorder, :start_server, true)
    Application.put_env(:tricorder, :grpc_port, port)
    Application.put_env(:tricorder, :workspace_root, workspace_root)
    :ok = Application.stop(:tricorder)
    {:ok, _} = Application.ensure_all_started(:tricorder)

    Tricorder.Deps.scan()

    loop()
  end

  defp loop(), do: loop()
end
