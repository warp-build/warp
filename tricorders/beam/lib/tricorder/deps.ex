defmodule Tricorder.Deps do
  def scan() do
    GenServer.cast(Tricorder.Deps.Server, :scan)
  end

  def loaded?() do
    GenServer.call(Tricorder.Deps.Server, :loaded?, :infinity)
  end
end
