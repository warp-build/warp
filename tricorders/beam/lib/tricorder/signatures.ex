defmodule Tricorder.Signatures do
  def mix_escript(name, file) do
    %{
      name: file,
      bin: name,
      rule: "mix_escript"
    }
  end

  def rebar3_library(name, file) do
    %{
      name: file,
      lib: name,
      rule: "rebar3_library"
    }
  end

  def elixir_library(file) do
    %{
      name: file,
      rule: "elixir_library",
      srcs: [Path.basename(file)]
    }
  end

  def elixir_script(file) do
    %{
      name: file,
      rule: "elixir_script",
      main: Path.basename(file)
    }
  end

  def erlang_library(file, modules, includes) do
    %{
      name: file,
      rule: "erlang_library",
      srcs: [Path.basename(file)],
      includes: includes,
      modules: modules
    }
  end

  def erlang_header_library(file, include_deps) do
    %{
      name: file,
      rule: "erlang_library",
      srcs: [Path.basename(file)],
      includes: include_deps
    }
  end

  def erlang_test(file, case_name, modules, includes) do
    %{
      name: case_name,
      rule: "erlang_test",
      test: file,
      modules: modules,
      includes: includes,
      cases: [case_name]
    }
  end
end
