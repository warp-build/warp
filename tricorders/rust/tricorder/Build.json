{
    "signatures": [
	{
	    "name": "tricorder",
	    "rule": "cargo_binary",
	    "bin": "tricorder",
	    "srcs": ["**/*"]
	}
    ]
}
