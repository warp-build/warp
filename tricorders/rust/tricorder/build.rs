fn main() {
    tonic_build::configure()
        .build_server(true)
        .include_file("_include.rs")
        .compile_well_known_types(true)
        .compile(
            &[
                "../../../schemas/build/warp/codedb/analyzer.proto",
                "../../../schemas/build/warp/resolver.proto",
                "../../../schemas/build/warp/common.proto",
            ],
            &["../../.."],
        )
        .unwrap_or_else(|e| panic!("protobuf compile error: {}", e));
}
