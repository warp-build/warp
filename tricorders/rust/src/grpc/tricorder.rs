use super::mappers::ast_to_success_response;
use super::proto::build::warp::tricorder::tricorder_service_server::{
    TricorderService, TricorderServiceServer,
};
use super::proto::build::warp::tricorder::{
    generate_signature_response, EnsureReadyRequest, EnsureReadyResponse, GenerateSignatureRequest,
    GenerateSignatureResponse, GenerateSignatureSuccessResponse, GetAstRequest, GetAstResponse,
    PrepareDependencyRequest, PrepareDependencyResponse,
};
use super::GrpcTricorderError;
use crate::analysis::model::{Ast, Symbol};
use crate::analysis::Analyzer;
use crate::dependencies::DependencyManager;
use std::net::SocketAddr;
use std::path::Path;
use std::sync::Arc;
use tonic::transport::Server;
use tonic::{Request, Response, Status};

#[derive(Builder)]
#[builder(build_fn(error = "GrpcTricorderError"))]
pub struct GrpcTricorder {
    address: SocketAddr,
    dep_manager: Arc<DependencyManager>,
    analyzer: Arc<Analyzer>,
}

impl GrpcTricorder {
    pub fn builder() -> GrpcTricorderBuilder {
        Default::default()
    }

    pub async fn serve(self) -> Result<(), GrpcTricorderError> {
        let address = self.address;

        Server::builder()
            .add_service(TricorderServiceServer::new(self))
            .serve(address)
            .await?;

        Ok(())
    }
}

#[tonic::async_trait]
impl TricorderService for GrpcTricorder {
    async fn ensure_ready(
        &self,
        _request: Request<EnsureReadyRequest>,
    ) -> Result<Response<EnsureReadyResponse>, Status> {
        self.dep_manager.wait_until_ready().await;
        Ok(Response::new(EnsureReadyResponse::default()))
    }

    async fn generate_signature(
        &self,
        request: Request<GenerateSignatureRequest>,
    ) -> Result<Response<GenerateSignatureResponse>, Status> {
        let request_data = request.into_inner();
        let file = request_data.clone().file;
        let test_matcher: Vec<String> = match request_data.test_matcher {
            Some(matches) => matches.into(),
            None => Vec::new(),
        };
        let workspace_root = request_data.workspace_root;

        let response = self
            .analyzer
            .generate_signature(workspace_root.clone(), Path::new(&file), test_matcher)
            .await;

        match response {
            Ok(signature) => Ok(Response::new(GenerateSignatureResponse {
                response: Some(generate_signature_response::Response::Ok(
                    GenerateSignatureSuccessResponse {
                        workspace_root,
                        file,
                        signatures: signature.iter().map(|e| e.clone().into()).collect(),
                    },
                )),
            })),
            // TODO(@calin): create a proper way of dealing with missing dependencies.
            Err(_) => Ok(Response::new(GenerateSignatureResponse {
                response: Some(generate_signature_response::Response::Ok(
                    GenerateSignatureSuccessResponse {
                        workspace_root,
                        file,
                        signatures: vec![],
                    },
                )),
            })),
        }
    }

    async fn get_ast(
        &self,
        request: Request<GetAstRequest>,
    ) -> Result<Response<GetAstResponse>, Status> {
        let request = request.into_inner();

        let sym = match request.test_matcher {
            Some(test_matcher) if !test_matcher.raw.is_empty() => Symbol::Named {
                name: test_matcher.raw[0].to_string(),
            },
            _ => Symbol::All,
        };

        let response = self.analyzer.get_ast(request.file, sym).await;

        match response {
            Ok(ast) => Ok(Response::new(ast_to_success_response(ast))),
            // TODO(@calin): create a proper Requirement internal model for missing deps
            // and return it properly here.
            Err(_) => Ok(Response::new(ast_to_success_response(Ast::default()))),
        }
    }

    async fn prepare_dependency(
        &self,
        _request: Request<PrepareDependencyRequest>,
    ) -> Result<Response<PrepareDependencyResponse>, Status> {
        Ok(Response::new(PrepareDependencyResponse::default()))
    }
}
