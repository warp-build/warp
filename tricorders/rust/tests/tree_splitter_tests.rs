use tricorder::{Symbol, TreeSplitter};

#[test]
fn t001_simple_fn_graph() {
    let src = include_str!("./sources/t001_simple_fn_graph.rs");
    let symbol = Symbol::builder()
        .name("main".to_string())
        .scope(tricorder::SymbolScope::Named)
        .build()
        .unwrap();

    let (ast, filtered_source) = TreeSplitter::tree_split(symbol, src);
    insta::assert_debug_snapshot!(ast);
    insta::assert_display_snapshot!(format!("{}", &filtered_source));
}
