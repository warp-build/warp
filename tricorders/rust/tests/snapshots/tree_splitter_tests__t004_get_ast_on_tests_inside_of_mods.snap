---
source: tricorders/rust/tests/tree_splitter_tests.rs
expression: deps
---
[
    "add_two",
    "assert_failed",
]
